exchange/bidder_test.go

```diff
@@ -0,0 +1,412 @@
+	}
+	if ext.RequestBody != string(info.request.Body) {
+		t.Errorf("The request body should be empty. Got %s", ext.RequestBody)
+	}
+	if ext.ResponseBody != "" {
+		t.Errorf("The response body should be empty. Got %s", ext.ResponseBody)
+	}
+	if ext.Status != 0 {
+		t.Errorf("The Status code should be 0. Got %d", ext.Status)
+	}
+}
+
+// TestSuccessfulResponseLogging makes sure that openrtb_ext works properly if the HTTP request is successful.
+func TestSuccessfulResponseLogging(t *testing.T) {
+	info := &httpCallInfo{
+		request: &adapters.RequestData{
+			Uri:  "test.com",
+			Body: []byte("request body"),
+		},
+		response: &adapters.ResponseData{
+			StatusCode: 200,
+			Body:       []byte("response body"),
+		},
+	}
+	ext := makeExt(info)
+	if ext.Uri != info.request.Uri {
+		t.Errorf("The URI should be test.com. Got %s", ext.Uri)
+	}
+	if ext.RequestBody != string(info.request.Body) {
+		t.Errorf("The request body should be \"request body\". Got %s", ext.RequestBody)
+	}
+	if ext.ResponseBody != string(info.response.Body) {
+		t.Errorf("The response body should be \"response body\". Got %s", ext.ResponseBody)
+	}
+	if ext.Status != info.response.StatusCode {
+		t.Errorf("The Status code should be 0. Got %d", ext.Status)
+	}
+}
+
+// TestServerCallDebugging makes sure that we log the server calls made by the Bidder on test bids.
+func TestServerCallDebugging(t *testing.T) {
+	respBody := "{\"bid\":false}"
+	respStatus := 200
+	server := httptest.NewServer(mockHandler(respStatus, "getBody", respBody))
+	defer server.Close()
+
+	reqBody := "{\"key\":\"val\"}"
+	reqUrl := server.URL
+	bidderImpl := &goodSingleBidder{
+		httpRequest: &adapters.RequestData{
+			Method:  "POST",
+			Uri:     reqUrl,
+			Body:    []byte(reqBody),
+			Headers: http.Header{},
+		},
+	}
+	bidder := adaptBidder(bidderImpl, server.Client())
+
+	bids, _ := bidder.requestBid(context.Background(), &openrtb.BidRequest{
+		Test: 1,
+	})
+
+	if len(bids.httpCalls) != 1 {
+		t.Errorf("We should log the server call if this is a test bid. Got %d", len(bids.httpCalls))
+	}
+	if bids.httpCalls[0].Uri != reqUrl {
+		t.Errorf("Wrong httpcalls URI. Expected %s, got %s", reqUrl, bids.httpCalls[0].Uri)
+	}
+	if bids.httpCalls[0].RequestBody != reqBody {
+		t.Errorf("Wrong httpcalls RequestBody. Expected %s, got %s", reqBody, bids.httpCalls[0].RequestBody)
+	}
+	if bids.httpCalls[0].ResponseBody != respBody {
+		t.Errorf("Wrong httpcalls ResponseBody. Expected %s, got %s", respBody, bids.httpCalls[0].ResponseBody)
+	}
+	if bids.httpCalls[0].Status != respStatus {
+		t.Errorf("Wrong httpcalls Status. Expected %d, got %d", respStatus, bids.httpCalls[0].Status)
+	}
+}
+
+func TestErrorReporting(t *testing.T) {
+	bidder := adaptBidder(&bidRejector{}, nil)
+	bids, errs := bidder.requestBid(context.Background(), &openrtb.BidRequest{})
+	if bids != nil {
+		t.Errorf("There should be no seatbid if no http requests are returned.")
+	}
+	if len(errs) != 1 {
```

Comment:
Could also check the contents of the error received to make sure it's what is expected.